package/ota_tool/src/main.rs

```diff
@@ -0,0 +1,160 @@
+#[macro_use]
+extern crate hyper;
+extern crate reqwest;
+extern crate uuid;
+#[macro_use]
+extern crate serde_derive;
+extern crate clap;
+
+use uuid::Uuid;
+use std::process::Command;
+use std::io;
+use std::io::{BufReader, Write};
+use std::fs::File;
+use std::path::Path;
+use std::iter::Iterator;
+use clap::{Arg, App};
+
+header! { (DeviceUid, "Device-Uid") => [Uuid] }
+header! { (CurrentVersion, "Current-Version") => [String] }
+
+#[derive(Deserialize, Debug)]
+struct OTAServiceResponse {
+    version: String,
+    sha256: String,
+    url: String,
+}
+
+fn sha256(path: &std::path::Path) -> Result<String, String> {
+    // Linux
+    let mut base_cmd = Command::new("sha256sum");
+    // OS X:
+    // let mut base_cmd = Command::new("shasum");
+    //         base_cmd.args(&["-a", "256"]);
+    let hash_cmd_output = String::from_utf8(
+                            base_cmd
+                            .arg(path.to_str().ok_or("Path error")?)
```

Comment:
I find the `?` syntax pretty problematic since it de-sugars to an early return.  It's probably the python programmer in me, but I feel it's useful when the visual layout of the code can communicate some of the control flow.  Thus something like this would seem better in that regard:

```rust
fn sha256(path: &std::path::Path) -> Result<String, String> {

    let mut base_cmd = Command::new("sha256sum");

    let path_s = path
        .to_str()
        .ok_or("Path decode error")
        ?;

    let cmd_result = base_cmd
        .arg(path_s)
        .output()
        .map_err(|e| format!("Error calling shasum: {}", e.to_string()))
        ?;

    let hash_cmd_output = String::from_utf8(cmd_result.stdout)
        .map_err(|e| format!("Error encoding shasum ouput: {}", e.to_string()))
        ?;

    let output_split = hash_cmd_output
        .split(" ")
        .next()
        .ok_or("Error parsing shasum output")
        ?;

    return Ok(String::from(output_split));
}
```